7 - feat(android): support notifyAppReady() in manual mode

lincolnthree opened the pull request:
Second attempt at implementing: https://github.com/Cap-go/capacitor-updater/issues/60

This PR is much more finished than my prior attempt, and has also resolved a few additional issues:

1. Calling `reset()` now properly reloads the webview an can revert to 'public' builtin version as expected.
2. Adds `next()` method, which simply sets the next version to applied on app restart (like autoUpdate does)
3. Adds `versionName` to `download()`, to avoid requiring call to `set()` to save the version name info.
4. Plugin methods now return `VersionInfo` where appropriate, which includes `{ version: string, versionName: string, downloaded: string | Date, status: VersionStatus }`.
5. Re-designed how version updating, `VersionStatus` success/failure states are implemented. All code (manual and autoUpdate) both use this new system.
6. Adds type-safe `PluginsConfig` for capacitor.config.ts
7. Adds `autoDeleteFailed` boolean config to enable/disable automatic deletion of failed updates
8. Adds `autoDeletePrevious` boolean config to enable/disable automatic deletion of prior updates (upon successful update) 
9. Adds configurable `appReadyTimeout` int config to control how long the native plugin will wait before considering a bundle 'failed' 
10. Adds additional exception handling to most API calls, to prevent any native crashes we may have missed. (Better to be paranoid than crash for users, right?)
11. Adds some additional configuration parameters for capacitor.config.ts (see `definitions.ts`)


Looking forward to your feedback!

vorokami commented:
@lincolnthree Hi, Checked your code and found thing that don't will be working properly. Capacitor webview using preference "serverBasePath" (see link below). And if we dont change this value in this prefs
SharedPreferences prefs = getContext().getSharedPreferences("CapWebViewSettings", Activity.MODE_PRIVATE);
then on restart, application will be start builtin version instead of verstion that was set. And in your code i dont found this.

https://github.com/ionic-team/capacitor/blob/b893a57aaaf3a16e13db9c33037a12f1a5ac92e0/android/capacitor/src/main/java/com/getcapacitor/plugin/WebView.java#L15

In this plugin (Cap-go/capacitor-updater) it is used (here links to where it is used specifically)
https://github.com/Cap-go/capacitor-updater/blob/dc578b591fdadb446947b0a747b5fe4a5ad444a4/android/src/main/java/ee/forgr/capacitor_updater/CapacitorUpdater.java#L290
and
https://github.com/Cap-go/capacitor-updater/blob/dc578b591fdadb446947b0a747b5fe4a5ad444a4/android/src/main/java/ee/forgr/capacitor_updater/CapacitorUpdater.java#L355

lincolnthree commented:
@vorokami Thanks for pointing this out. It definitely wasn't clear on the original code that this was necessary! 

I've updated my patch to use the type-safe references to the `WebView` class static constants to make this clear, and also removed the unnecessary/redundant `bundlePath` variable, which now simply uses `serverBasePath` as, so far as I understand, this should be the sole source of truth.

Let me know your thoughts!

vorokami commented:
Yes, I also think this is the best way to use the `serverBasePath`.
